integration/integration_linux_test.go

```diff
@@ -225,23 +226,29 @@ func (n Namespace) Del() {
 }
 
 func makeTcpClientInNS(netns string, address string, port int, numBytes int) {
-	message := bytes.Repeat([]byte{'a'}, numBytes)
+	message := bytes.Repeat([]byte{'a'}, numBytes+1)
+	message[numBytes] = '\n' // sending final \n so we are sure nc sends the input
 
 	bin, err := exec.LookPath("nc")
 	Expect(err).NotTo(HaveOccurred())
 	var cmd *exec.Cmd
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
 	if netns != "" {
 		netns = filepath.Base(netns)
-		cmd = exec.Command("ip", "netns", "exec", netns, bin, "-v", address, strconv.Itoa(port))
+		cmd = exec.CommandContext(ctx, "ip", "netns", "exec", netns, bin, "-v", address, strconv.Itoa(port))
 	} else {
-		cmd = exec.Command("nc", address, strconv.Itoa(port))
+		cmd = exec.CommandContext(ctx, bin, address, strconv.Itoa(port))
 	}
 	cmd.Stdin = bytes.NewBuffer([]byte(message))
 	cmd.Stderr = GinkgoWriter
 	out, err := cmd.Output()
+	Expect(err).Should(Or(Not(HaveOccurred()),
+		ContainSubstring("signal: killed"),
```

Comment:
what's going on here?  the test passes if the subprocess is killed?

Reply:
When I changed ubuntu version, all the tests using `nc` started to hang causing a major headache. My understanding is that, as per https://unix.stackexchange.com/questions/332163/netcat-send-text-to-echo-service-read-reply-then-exit , the bump in ubuntu resulted in a different nc version, which hangs waiting for new input while the previous did not. I could have added the -q parameter, making it incompatible with the OpenBSD version.

To make this test portable, and work on both versions here I use exec.CommandContext which kills the process if the timeout on the context expires. As a side effect, having the process still running waiting for extra input but killed for timeout is not an error condition.

The success or failure of the test is validated in the following lines, where we check for the output.

Reply:
@bboreham does it make sense? Not sure if there was a more elegant way to do so.

Reply:
Does it mean the test takes a minute longer?

Reply:
It's been a while, but yes. If nc hangs, we wait for the timeout on the context to check the output.
I can make it way shorter though, 2 seconds should be more than enough.

Reply:
We discussed this at the maintainers meeting.  It's way more convoluted than I realised.

The test where this is used [specifically checks the time taken](https://github.com/containernetworking/plugins/blob/e78e6aa5b9fd7e3e66f0cb997152c44c2a4e43df/integration/integration_linux_test.go#L188), so any approach with a timeout is unwise.

Looking at [the nc man page](https://www.commandlinux.com/man-page/man1/nc.traditional.1.html), it should exit when the _network_ side closes. In this test the network side is an echo server which apparently does [close the connection](https://github.com/containernetworking/plugins/blob/e78e6aa5b9fd7e3e66f0cb997152c44c2a4e43df/pkg/testutils/echosvr/main.go#L52), so perhaps you could investigate why that isn't enough?

Reply:
Sure, will do and get back.
Thanks for now!